### ISSUE TITLE
webassembly scheduling of JavaScript callbacks is not fair

### ISSUE THREAD

**Issue by sherhut** [golang/go#27894]
### What version of Go are you using (`go version`)?

go version go1.11 linux/amd64

### Does this issue reproduce with the latest release?

yes

### Description

I ran the webassembly benchmark in https://djhworld.github.io/post/2018/09/21/i-ported-my-gameboy-color-emulator-to-webassembly/ and as also noted on that page, the benchmark does not properly process keyboard input on chrome. See https://bugs.chromium.org/p/v8/issues/detail?id=8227 for the v8 tracking bug.

Due to a performance issue in v8, the compiled go code runs slower in v8 than firefox. As a consequence, there is more pressure on the go scheduler. In particular, there is nearly always a go-routine ready to be executed with very few idle pauses. Looking at the scheduler implementation at

https://github.com/golang/go/blob/eae5fc88c13d6902a7fe9c595fb02155eb037cbe/src/runtime/proc.go#L2368

this leads to starvation of the go-routines that are waiting for a callback from java-script. In case of the gameboy emulator, this means that no more I/O events are being processed.

**Comment by myitcv** [comment]
This is in some way related to https://github.com/golang/go/issues/26045?

**Comment by agnivade** [comment]
Possibly. But I am not very clear on what exactly this issue tries to resolve. Does `pauseSchedulerUntilCallback` performs a STW-like action where unless the callback hasn't happened, no other goroutine can execute ?

Otherwise, I don't see a reason other goroutines will still wait for a callback from js.

**Comment by sherhut** [comment]
The issue I am seeing here is not that go-routines cannot run while we wait for a callback, but that callbacks do not happen until no other go-routine runs. I just briefly looked at the runtime and might have missed something. But this is how I understand it:

Go (running on top of WebAssembly) can only receive callbacks from JavaScript if the WebAssembly runtime returns to the event loop. This is facilitated by pauseSchedulerUntilCallback, which in the go-runtime for JavaScript calls pause if any callbacks from JavaScript are potentially pending, that is a callback handler has been registered before. I looked at

https://github.com/golang/go/blob/50bd1c4d4eb4fac8ddeb5f063c099daccfb71b26/src/runtime/lock_js.go#L148

The call to pause will fully unwind the stack and return from WebAssembly back to JavaScript land.

https://github.com/golang/go/blob/50bd1c4d4eb4fac8ddeb5f063c099daccfb71b26/src/runtime/rt0_js_wasm.s#L72

Is there another path to fully unwind and return to JavaScript that I have missed?

**Comment by neelance** [comment]
Yes, giving back execution from Go (WebAssembly) to JavaScript is a "stop the world" action for Go. No goroutine will run until a callback invokes Go code (user action, timer, etc.). Since we want all Go code to run which is not waiting for such callbacks, we give back execution to JS only if ALL goroutines are asleep. If your Go code always has something to do, execution will never be passed back to JS, so user events will not be processed.

This is the current situation. Maybe we could extend the scheduler so it periodically gives back execution to JS, even if there are still runnable goroutines.

This somewhat depends on #26045.

**Comment by agnivade** [comment]
@neelance - Just wanted to check in on this. Now that #26045 is resolved, has this situation improved, or it still needs further work ?

**Comment by neelance** [comment]
I think this needs further work. Most likely it will get resolved when we add WebAssembly threads.

**Comment by semaj** [comment]
@neelance I think I'm running into this same issue, but I'm a bit confused by your explanation. How are goroutines scheduled in the browser environment? If there are 3 pending goroutines, does this equate to 3 pending events in the event queue? If it does, shouldn't an I/O event simply take its place in the event queue after the pending goroutines?

**Comment by neelance** [comment]
No, goroutines are not mapped to the event queue. If there are still goroutines awake, then the event queue will not get processed.

**Comment by semaj** [comment]
Thanks for responding! 

What constitutes "awake"? If one goroutine is blocked on a channel read and another is sleeping via `time.Sleep`, will Go immediately relinquish control to JavaScript to process pending events?
